keanu-project/src/main/java/io/improbable/keanu/vertices/bool/BoolVertex.java

```diff
@@ -91,6 +93,24 @@ public boolean getValue(int... index) {
         return getValue().getValue(index);
     }
 
+    public BooleanTensor sampleScalarValuesAsTensor(int[] shape, KeanuRandom random) {
```

Comment:
I think this could be a parameterised method on `Vertex`: 

    public T sample(int[] shape, KeanuRandom random)

Reply:
That's the road I went down initially, but things got hairy pretty quickly as the interface to create tensors uses arrays. This is primarily due to generics requiring boxed primitives, but the tensor constructors requiring unboxed primitives.

Finally, as this thing returns a tensor, it doesn't make sense to exist on a vertex that doesn't expose something that can be stored in a tensor (e.g. an enum or a user-defined object).

Richer type systems would let us solve these problems but this is the world we live in.

Reply:
If there's a way we can have this as a generic function regardless of these issues, I'd love to hear it (this duplicated code isn't ideal)

Reply:
We can probably have a go at removing a lot of the shared code here. What we're essentially doing is taking many samples and then reshaping them into a tensor. We can have a method on Vertex that knew how to take many samples and apply a function (in this case to go from tensor to scalar).

```
    public <U> List<U> sampleMany(int n, KeanuRandom random, Function<T, U> func) {
        ArrayList<U> samples = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            samples.add(func.apply(sample(random)));
        }
        return samples;
    }
```

Then we can start doing:

```
    public BooleanTensor sampleScalarValuesAsTensor(int[] shape, KeanuRandom random) {
        if (!TensorShape.isScalar(this.getShape())) {
            throw new IllegalArgumentException("Vertex to sample must be scalar");
        }

        final int length = Math.toIntExact(TensorShape.getLength(shape));
        List<Boolean> samples = sampleMany(length, random, i -> i.scalar());
        return BooleanTensor.create(Booleans.toArray(samples), shape);
    }
```

Reply:
I think that's a great idea.

I'd probably make a new class for this, rather than putting it into `Vertex`. Also (optional) it could take a `Vertex` rather than a `Function<T, U>` - that way you can put the shape checking into the method too, rather than repeat it 3 times.

    package io.improbable.keanu.vertices;
    
    import java.util.ArrayList;
    import java.util.List;
    
    import io.improbable.keanu.tensor.Tensor;
    import io.improbable.keanu.tensor.TensorShape;
    import io.improbable.keanu.vertices.dbl.KeanuRandom;
    
    public class ScalarSampler<U> {
    
        public List<U> sample(Vertex<? extends Tensor<U>> vertex, int[] shape, KeanuRandom random) {
            if (!TensorShape.isScalar(vertex.getShape())) {
                throw new IllegalArgumentException("Vertex to sample must be scalar");
            }
    
            final int length = Math.toIntExact(TensorShape.getLength(shape));
            ArrayList<U> samples = new ArrayList<>();
            for (int i = 0; i < length; i += 1) {
                samples.add(vertex.sample(random).scalar());
            }
            return samples;
        }
    }